vtpm/src/socket.rs

```diff
@@ -0,0 +1,192 @@
+    // Control Channel used by Cloud-Hypervisor
+    ctrl_fd: RawFd,
+}
+
+impl Default for SocketDev {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl SocketDev {
+    pub fn new() -> Self {
+        Self {
+            state: SocDevState::Disconnected,
+            stream: None,
+            write_msgfd: -1,
+            ctrl_fd: -1,
+            data_fd: -1,
+        }
+    }
+
+    pub fn init(&mut self, path: String) -> Result<()> {
+        self.connect(&path)?;
+        Ok(())
+    }
+
+    pub fn connect(&mut self, socket_path: &str) -> Result<()> {
+        self.state = SocDevState::Connecting;
+
+        let now = Instant::now();
+
+        // Retry connecting for 10 secs
+        loop {
+            match UnixStream::connect(socket_path) {
+                Ok(s) => {
+                    let fd = s.as_raw_fd();
+                    self.ctrl_fd = fd;
+                    self.stream = Some(s);
+                    self.state = SocDevState::Connected;
+                    debug!("Connected to vTPM socket path : {:?}", socket_path);
+                    return Ok(());
+                }
+                Err(_e) => {}
+            };
+            sleep(Duration::from_millis(1000));
+
+            if now.elapsed().as_secs() >= 10 {
+                break;
+            }
+        }
+        Err(TPMSocError::ConnectToTPMSocket(anyhow!(
+            "Failed to connect to vTPM Socket"
+        )))
+    }
+
+    pub fn set_datafd(&mut self, fd: RawFd) {
+        self.data_fd = fd;
+    }
+
+    pub fn set_msgfd(&mut self, fd: RawFd) {
+        self.write_msgfd = fd;
+    }
+
+    pub fn send_full(&self, buf: &mut [u8], len: usize) -> Result<usize> {
+        let write_fd = self.write_msgfd;
+        let fds = &[write_fd];
+
+        let cmsg_buf = Vec::<Vec<u8>>::with_capacity(unsafe { CMSG_SPACE(1) } as usize);
+
+        let data_vecs = [libc::iovec {
+            iov_base: buf.as_mut_ptr() as *mut libc::c_void,
+            iov_len: len,
+        }; 1];
+
+        let mut msghdr: libc::msghdr = unsafe { zeroed() };
+        msghdr.msg_name = ptr::null_mut();
+        msghdr.msg_namelen = 0;
+        msghdr.msg_iov = data_vecs.as_ptr() as *mut libc::iovec;
+        msghdr.msg_iovlen = data_vecs.len();
+
+        //Append one Ancillary Message with a single FD
+        msghdr.msg_controllen = unsafe { CMSG_SPACE(1) } as usize;
+        msghdr.msg_control = cmsg_buf.as_ptr() as *mut libc::c_void;
+
+        unsafe {
+            let cmsghdr = CMSG_FIRSTHDR(&msghdr);
+            (*cmsghdr).cmsg_level = SOL_SOCKET;
+            (*cmsghdr).cmsg_type = SCM_RIGHTS;
+            (*cmsghdr).cmsg_len = CMSG_LEN(mem::size_of::<i32>() as u32) as usize;
+            let fdptr = CMSG_DATA(cmsghdr);
+            ptr::copy_nonoverlapping(fds as *const _ as *const u8, fdptr, mem::size_of::<i32>());
+        }
+
+        // Send Ancillary data, along with cmds and data
+        let ret = unsafe { libc::sendmsg(self.ctrl_fd, &msghdr, 0) };
```

Comment:
Try this: `vmm_sys_util::sock_ctrl_msg::ScmSocket` for sending FDs over a unix socket.